generalize the portable MF methods to fe_degree < n_q_points_1d cases

YiminJin opened the pull request:
This PR is identical to #17957 . #17957 is too far behind the current master branch and is too hard to rebase, so I decide to insert the modifications to the current master branch manually and submit a new PR. 

I am aware of #18039 that generalizes the portable MF functions to multi-component cases, which modifies the same files as # 17957 does. I have checked the conflicts and made some modifications on the basis of #18039. I have tested the multi-component case with fe_degree < n_q_points_1d and the results are correct.

tjhei commented:
> This PR is identical to #17957 . #17957 is too far behind the current master branch and is too hard to rebase

One thing you can do is:
1. squash all of 17957 into a single commit (by rebasing with -i to its old parent commit)
2. try again to rebase.

YiminJin commented:
> squash all of 17957 into a single commit (by rebasing with -i to its old parent commit)

@tjhei I do not quite understand the operation. What I have tried is:
1. `git rebase -i master`
2. replace "pick" by "squash" except for the old parent commit in file git-rebase-todo (there are 233 commits to squash.  I guess the reason for this is that I did something wrong on Monday when trying to rebase to the master branch)

After that, git keeps telling me that there are conflicts, and I have to deal with all the conflicts manually. I struggled for two hours yesterday and finally gave up because it was too easy to make mistakes.

I have also tried `git rebase -i hash` where `hash` is my old parent commit, and the result is similar --- over 200 commits to squash and a lot of conflicts. Do you think there are any remedies?

Anyway I should apologize for the mistakes. I knew `git rebase` is sometimes dangerous and I should have been more careful.

YiminJin commented:
@masterleinad Thank you for the error report. I think the errors are caused by `AssertThrow()` in `FEEvaluation::evaluate()` and `FEEvaluation::integrate()` which is illegal in device code, and I have replaced it by `Kokkos::abort()` in a new commit.

However, the errors reported by the cuda-11 clang checker are not fixed. I think those errors are caused by line 77 in tests/tests.h:
```#define Assert AssertThrow```
Since `AssertThrow()` is not defined for device code, all the `Assert()` functions in include/dealii/matrix_free/portable_tensor_product_kernels.h become illegal. An easy solution is to replace `Assert(...)` by 
``` 
#ifdef DEBUG
KOKKOS_ASSERT(...)
#endif
 ```
Do you think this is appropriate, or shall I remove the assertions temporarily?

tjhei commented:
> I do not quite understand the operation. What I have tried is:

The problem is that you have done a combination of merges from master and adding additional work/fixes on top. My general advice for the future would be:
1. never use ``git pull`` or ``git merge`` to bring in a newer version.
2. squash your commits to a single commit (git rebase but choosing the commit you started on, not the current master)
3. rebase to current master often
4. try to make many small PRs and get them merged instead of a single large PR that is open for a long time.

I know this is annoying. Sorry. :-(

tjhei commented:
I forgot to say:
Here is your old branch squashed to a single commit: https://github.com/dealii/dealii/compare/master...tjhei:dealii:mf-degree-split

I did this by doing:
```
git checkout master
git pull upstream master
git fetch refs/pull/17957/head
git merge --no-commit --squash FETCH_HEAD
```

masterleinad commented:
My 2c, I like using `git merge` (since that requires less merge conlfict resolutions) but for squashing would do
- git merge upstream/master
- git reset upstream/master
- git add (all changes)
- git commit

masterleinad commented:
> However, the errors reported by the cuda-11 clang checker are not fixed. I think those errors are caused by line 77 in tests/tests.h:
#define Assert AssertThrow
Since AssertThrow() is not defined for device code, all the Assert() functions in include/dealii/matrix_free/portable_tensor_product_kernels.h become illegal. An easy solution is to replace Assert(...) by
[...]

I would just disable https://github.com/dealii/dealii/blob/b556c95dd0282ea07818c76c29f02c47886227d0/tests/tests.h#L72-L77 if any Kokkos GPU backend is enabled using
```
#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP) || defined(KOKKOS_ENABLE_SYCL)
```

YiminJin commented:
@tjhei @masterleinad Thank you very much! There are still some commands that I do not understand, but I will do the rebase in a safer way than last time.

I have compared the changed files in https://github.com/dealii/dealii/compare/master...tjhei:dealii:mf-degree-split with the present PR, and they are almost identical (except for a small bug that has been fixed in this PR).

YiminJin commented:
> I would just disable ... if any Kokkos GPU backend is enabled using

I have tried this in a new commit. This is a little dangerous because I cannot test the modifications with a clang complier on my machine. Besides, my modifications have also disabled the redefinition in host codes when Kokkos GPU backends are enabled, which is not what we want. I just hope the modification could satisfy the clang checker.

masterleinad commented:
/rebuild

tjhei commented:
Test is still failing: https://ci.tjhei.info/job/dealii-serial/job/PR-18124/5/testReport/projectroot.tests.matrix_free_kokkos.evaluate_2d_shapedebug/serial/matrix_free_kokkos_evaluate_2d_shape_debug/

tjhei commented on include/deal.II/matrix_free/portable_fe_evaluation.h:
```diff
@@ -340,50 +344,40 @@ namespace Portable
+      }
+    else if (fe_degree >= 0 &&
+             internal::use_collocation_evaluation(fe_degree, n_q_points_1d) &&
+             data->element_type <= ElementType::tensor_symmetric)
```
Could you leave a comment what "<=" means here (and further below)?

YiminJin replied:
I have copied the comments in `internal::FEEvaluationImplSelector::run()` to the same place here.

tjhei commented on include/deal.II/matrix_free/portable_fe_evaluation.h:
```diff
@@ -512,17 +498,17 @@ namespace Portable
                                             Number>::value_type
   FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
-    get_dof_value(int q_point) const
+    get_dof_value(int dof) const
```
can you add an assert to make sure this index is valid? (or does that happen insides ``values``?)

YiminJin replied:
I have added assert to all the functions that take an index as argument, as in class `dealii::FEEvaluation`.

tjhei commented on include/deal.II/matrix_free/portable_matrix_free.templates.h:
```diff
@@ -691,13 +744,15 @@ namespace Portable
-    Assert(n_dofs_1d == n_q_points_1d,
-           ExcMessage("n_q_points_1d must be equal to fe_degree+1."));
+    // TODO remove the limitation in the future
+    AssertThrow(n_dofs_1d <= n_q_points_1d,
```
why not use ``Assert`` here?

YiminJin replied:
I think this function is called only once per run (or per solving procedure), so using `AssertThrow` does not cause performance problem. Besides, the condition n_dofs_1d <= n_q_points_1d is easy to violate, since it is controlled by the template parameters specified by the users. So it might save some debugging time.

YiminJin commented:
I do not understand the error report of Jenkins: serial. It says that
```
Test matrix_free_kokkos/evaluate_2d_shape.debug: DIFF
===============================   OUTPUT BEGIN  ===============================
[ 25%] Building CXX object CMakeFiles/matrix_free_kokkos.evaluate_2d_shape.debug.dir/evaluate_2d_shape.cc.o
[ 50%] Linking CXX executable evaluate_2d_shape.debug/evaluate_2d_shape.debug
[ 50%] Built target matrix_free_kokkos.evaluate_2d_shape.debug
[ 75%] Normalizing test output file /home/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.debug/serial/output
[100%] Generating evaluate_2d_shape.debug/serial/diff
matrix_free_kokkos/evaluate_2d_shape.debug: BUILD successful.
matrix_free_kokkos/evaluate_2d_shape.debug: RUN successful.
matrix_free_kokkos/evaluate_2d_shape.debug: DIFF failed. ------ Source: /jenkins/workspace/dealii-serial_PR-18124/tests/matrix_free_kokkos/evaluate_2d_shape.output
matrix_free_kokkos/evaluate_2d_shape.debug: DIFF failed. ------ Result: /home/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.debug/serial/output
Check /home/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.debug/serial/output /jenkins/workspace/dealii-serial_PR-18124/tests/matrix_free_kokkos/evaluate_2d_shape.output
matrix_free_kokkos/evaluate_2d_shape.debug: DIFF failed. ------ Diff:   /home/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.debug/serial/failing_diff
matrix_free_kokkos/evaluate_2d_shape.debug: DIFF failed. ------ First 20 lines of numdiff output:
----------------
##3       #:7   <== 0
##3       #:7   ==> 0.383851
@ Absolute error = 3.8385100000e-1, Relative error = Inf
##3       #:8   <== 0
##3       #:8   ==> 0.270406
@ Absolute error = 2.7040600000e-1, Relative error = Inf
##3       #:9   <== 0
##3       #:9   ==> 0.101667
@ Absolute error = 1.0166700000e-1, Relative error = Inf
##3       #:11  <== 0
##3       #:11  ==> 0.0669184
@ Absolute error = 6.6918400000e-2, Relative error = Inf
##3       #:12  <== 0
##3       #:12  ==> 0.0432944
@ Absolute error = 4.3294400000e-2, Relative error = Inf
##3       #:13  <== 0
##3       #:13  ==> -0.0223006
@ Absolute error = 2.2300600000e-2, Relative error = Inf
##3       #:15  <== 0
```
However, the output of tests/matrix_free_kokkos/evaluate_2d_shape.cc on my machine is
```
JobId GEOIST Mon Feb 17 13:41:40 2025
DEAL:values::Test 4 x 4
DEAL:values::Errors no transpose: -2.22045e-16 5.55112e-17 0.00000 2.22045e-16 -2.77556e-17 -1.73472e-17 5.55112e-17 1.11022e-16 -5.89806e-17 2.08167e-17 -5.55112e-17 -1.31839e-16 -2.22045e-16 0.00000 -5.55112e-17 0.00000 
DEAL:values::Errors transpose:    -4.44089e-16 1.11022e-16 0.00000 -8.88178e-16 -2.22045e-16 6.93889e-17 8.32667e-17 -1.11022e-16 0.00000 -2.77556e-17 1.11022e-16 4.44089e-16 0.00000 0.00000 0.00000 0.00000 
DEAL:values::Test 3 x 3
DEAL:values::Errors no transpose: 0.00000 6.93889e-18 1.73472e-18 -1.38778e-17 -6.93889e-17 0.00000 3.46945e-18 -1.38778e-17 0.00000 
DEAL:values::Errors transpose:    0.00000 0.00000 -1.38778e-17 0.00000 0.00000 0.00000 3.46945e-18 0.00000 -1.04083e-17 
DEAL:gradients::Test 4 x 4
DEAL:gradients::Errors no transpose: 4.51028e-17 6.93889e-17 0.00000 -1.38778e-17 1.17961e-16 0.00000 1.11022e-16 -1.56125e-17 -3.65918e-17 -1.43115e-16 8.32667e-17 5.55112e-17 6.79524e-17 0.00000 5.55112e-17 8.32667e-17 
DEAL:gradients::Errors transpose:    0.00000 -5.55112e-17 -2.77556e-17 -5.55112e-17 1.38778e-17 0.00000 -2.77556e-17 4.16334e-17 0.00000 -5.55112e-17 0.00000 1.38778e-17 2.22045e-16 1.11022e-16 -2.77556e-17 2.77556e-17 
DEAL:gradients::Test 3 x 3
DEAL:gradients::Errors no transpose: 2.77556e-17 0.00000 1.90820e-17 -2.77556e-17 -2.42861e-17 0.00000 3.81639e-17 0.00000 0.00000 
DEAL:gradients::Errors transpose:    1.11022e-16 -1.11022e-16 -2.08167e-17 -8.32667e-17 -5.55112e-17 6.93889e-18 -1.66533e-16 0.00000 3.46945e-17 
DEAL:add:values::Test 4 x 4
DEAL:add:values::Errors no transpose: 2.22045e-16 -2.22045e-16 -4.44089e-16 0.00000 -2.22045e-16 -4.44089e-16 0.00000 1.11022e-16 2.22045e-16 0.00000 0.00000 2.22045e-16 0.00000 0.00000 -4.44089e-16 0.00000 
DEAL:add:values::Errors transpose:    4.44089e-16 2.22045e-16 2.22045e-16 -4.44089e-16 0.00000 -4.44089e-16 0.00000 -4.44089e-16 0.00000 0.00000 4.44089e-16 -2.22045e-16 4.44089e-16 0.00000 -4.44089e-16 0.00000 
DEAL:add:values::Test 3 x 3
DEAL:add:values::Errors no transpose: 0.00000 4.44089e-16 4.44089e-16 4.44089e-16 0.00000 0.00000 4.44089e-16 0.00000 -4.44089e-16 
DEAL:add:values::Errors transpose:    8.88178e-16 0.00000 0.00000 0.00000 0.00000 0.00000 -8.88178e-16 4.44089e-16 0.00000 
DEAL:add:gradients::Test 4 x 4
DEAL:add:gradients::Errors no transpose: 6.66134e-16 -8.88178e-16 -2.63678e-16 -2.22045e-16 4.44089e-16 0.00000 0.00000 -1.11022e-16 -2.77556e-16 0.00000 8.88178e-16 4.44089e-16 2.22045e-16 -1.66533e-16 -4.44089e-16 -4.44089e-16 
DEAL:add:gradients::Errors transpose:    -4.44089e-16 4.44089e-16 -4.44089e-16 2.22045e-16 2.22045e-16 -4.44089e-16 2.22045e-16 -2.22045e-16 8.88178e-16 -4.44089e-16 -4.44089e-16 -6.66134e-16 -2.22045e-16 0.00000 -4.44089e-16 -2.22045e-16 
DEAL:add:gradients::Test 3 x 3
DEAL:add:gradients::Errors no transpose: -2.22045e-16 -1.11022e-16 1.11022e-16 0.00000 -2.22045e-16 0.00000 0.00000 -2.22045e-16 0.00000 
DEAL:add:gradients::Errors transpose:    4.44089e-16 0.00000 4.44089e-16 0.00000 0.00000 0.00000 4.44089e-16 0.00000 -4.44089e-16 
```
The errors are very small. I do not understand why the absolute errors obtained by the checker are so large.

YiminJin commented:
@tjhei The check errrors have been cleaned. I have squashed the commits, but I did not know it would erase the records of my commits in this page. The error reported by the clang checker was caused by the incompatibility between tests/matrix_free_kokkos/evaluate_2d_shape.cc and the modified struct `internal::EvaluatorTensorProduct`. Please check file tests/matrix_free_kokkos/evaluate_2d_shape.cc for the newest modifications.

masterleinad commented on include/deal.II/matrix_free/portable_evaluation_kernels.h:
```diff
@@ -0,0 +1,620 @@
+                    eval.template gradients<0, false, true, false>(temp2, u);
+                  }
+              }
+            else
+              Kokkos::abort("dim must not exceed 3!");
```
Here was well.

masterleinad commented on include/deal.II/matrix_free/portable_evaluation_kernels.h:
```diff
@@ -0,0 +1,620 @@
+                eval.template co_gradients<0, false, true, false>(
+                  Kokkos::subview(grad_u, Kokkos::ALL, 0), u);
+              }
+            else
+              Kokkos::abort("dim must not exceed 3!");
```
and here.

masterleinad commented on include/deal.II/matrix_free/portable_evaluation_kernels.h:
```diff
@@ -0,0 +1,620 @@
+                    eval.template co_gradients<0, false, true, false>(
+                      Kokkos::subview(grad_u, Kokkos::ALL, 0), u);
+                  }
+                else
+                  Kokkos::abort("dim must not exceed 3!");
```
and here

masterleinad replied:
Depending on the implementation of `Kokkos::abort`, a `static_assert` should save some registers.

masterleinad commented on include/deal.II/matrix_free/portable_tensor_product_kernels.h:
```diff
@@ -474,407 +460,111 @@ namespace Portable
     {
-      apply<dim, n_q_points_1d, Number, direction, dof_to_quad, add, in_place>(
-        team_member, shape_values, in, out);
+      if (in_place)
```
```suggestion
      if constexpr (in_place)
```

masterleinad commented on include/deal.II/matrix_free/portable_tensor_product_kernels.h:
```diff
@@ -474,407 +460,111 @@ namespace Portable
     {
-      apply<dim, n_q_points_1d, Number, direction, dof_to_quad, add, in_place>(
-        team_member, shape_gradients, in, out);
+      if (in_place)
```
```suggestion
      if constexpr (in_place)
```

masterleinad commented on include/deal.II/matrix_free/portable_tensor_product_kernels.h:
```diff
@@ -474,407 +460,111 @@ namespace Portable
-          team_member.team_barrier();
-        }
-      else if constexpr (dim == 3)
+      if (in_place)
```
```suggestion
      if constexpr (in_place)
```

masterleinad commented on include/deal.II/matrix_free/portable_tensor_product_kernels.h:
```diff
@@ -42,16 +42,42 @@ namespace Portable
+      Assert(dst.size() >= N, ExcInternalError());
+      Assert(src.size() >= N, ExcInternalError());
+
+      Kokkos::parallel_for(Kokkos::TeamThreadRange(team_member, N),
```
```suggestion
      Kokkos::parallel_for(Kokkos::TeamVectorRange(team_member, N),
```

YiminJin replied:
The original author of the functions said that:
```
                             // FIXME check why using parallel_reduce
                             // ThreadVector is slower
```
I am not sure if `TeamThreadRange` should be changed here. In Kokkos lecture 4, the lecturer used a similar case to explain the usage of scratch space, and he chose `TeamThreadRange` rather than `TeamVectorRange` (page 38).

masterleinad replied:
This part doesn't expose any nested levels parallelisms so we should just through all the threads we have at it even if `vector_length=1` at the moment.
The code with the comment above has the opportunity for more levels of parallelism but we serialize since that seems to be better right now.

masterleinad replied:
Also, see slide 48 titled "Example: contractDataFieldScalar (10)".

YiminJin replied:
Thank you. I have noticed the usage of both `TeamThreadRange` and TeamVectorRange` in slide 47. It is late in Beijing time and I will work on it tomorrow, along with the tests.

YiminJin replied:
I understand it now. I will change `TeamThreadRange` to `TeamVectorRange` in the `apply_...` functions in the next commit.

tjhei commented:
> I have squashed the commits, but I did not know it would erase the records of my commits in this page. T

Yes, this is the whole point of squashing commits as it gives a cleaner and simpler history.

masterleinad commented on include/deal.II/matrix_free/portable_evaluation_kernels.h:
```diff
@@ -0,0 +1,620 @@
+                    !(integration_flag & EvaluationFlags::gradients))
+                  {
+                    eval.template values<0, false, false, false>(u, temp);
+                    copy_view_1d<false>(shared_data->team_member,
+                                        u,
+                                        temp,
+                                        fe_degree + 1);
```
Do the writes here need to be potentionally `atomic`?

YiminJin replied:
Since the function is defined in a .h file and can be used everywhere, I think there is risk of writing conflict. When the template argument is set to `true` (add), I have used `atomic_add` inside the function; when set to `false` (value assignment), I just write `dst(i) = src(i)`. Is there an atomic operator for value assignment? Actually I think writing conflict in value assignment is mistake in code logic and cannot be prevented.

masterleinad replied:
I think the logic is fine (and we need to use atomics when adding for continuous elements) but we don't necessarily copy. What about we change it to `populate_view`?

masterleinad replied:
I was simply suggesting to rename `copy_view_1d` since it sometimes adds all values for a particular index.

masterleinad reviewed (commented):
Would you mind adding some tests demonstrating that this allows using the framework with `fe_degree` < `n_q_points_1d`?

masterleinad reviewed (changes requested):
Please add some tests.

YiminJin commented:
Sorry I met some problems when trying to add a test case. I simply add `test<3, 4, 0, false>();` in the main function of tests/matrix_free_kokkos/evaluate_2d_shape.cc, and changed a few lines to make things correct. The test runs correctly on CPU, but fails on GPU. The error message is "invalid device function", which makes no sense to me. I need some more time to figure it out.

YiminJin commented:
I have met a problem when trying to add a test for `n_dofs_1d < n_q_points_1d` cases. The test is simple: test file tests/matrix_free_kokkos/evaluate_2d_shape.cc has already provided a template function 
```
template <int M, int N, int type, bool add>
test()
{ ... }
```
to test the tensor product evaluator in different situations, while only the cases with `M == N` are tested. So the easiest way for me to test `n_dofs_1d < n_q_points_1d` cases is to add `test<M, N, type, add>()` in the main function with `M < N`,  for instance `test<3, 4, 0, false>()` (there are also some trivial modifications to be made in function `evaluate_tensor_product()`). However, it turns out that the test runs correctly on CPU, but fails on GPU. The error message is: 
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaFuncGetAttributes(&attr, func) error( cudaErrorInvalidDeviceFunction): invalid device function /opt/trilinos/14.2.0-cuda-11.8/include/Cuda/Kokkos_Cuda_KernelLaunch.hpp:140
zsh: IOT instruction (core dumped)  ./evaluate_2d_shape
```
I learnt from CUDA documentation that the error message means 
[The requested device function does not exist or is not compiled for the proper device architecture.](https://docs.nvidia.com/cuda/archive/11.4.4/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1gg3f51e3575c2178246db0a94a430e0038b6af535e7e53d3f21e2437e8977b8c2e)
Yet the test code runs into the same error on both the workstation and my laptop, on which the Nvidia cards, gcc versions and CUDA versions are all different. 

I also tried to isolate the problematic part. It turns out that when `M` and `N` are hardcoded as constant integers instead of template arguments,  the test runs correctly; once `M` and `N` are treated as template arguments, the error occurs. The two versions of test codes are:
[evaluate_2d_shape_3x4_notemplate.cc.txt](https://github.com/user-attachments/files/18979498/evaluate_2d_shape_3x4_notemplate.cc.txt)
[evaluate_2d_shape_3x4.cc.txt](https://github.com/user-attachments/files/18979551/evaluate_2d_shape_3x4.cc.txt)

The problem is really annoying. Could anyone help me figure this out?

tjhei commented:
I am trying to understand your code, but I see the following:
```
template <int M, int N, int type, bool add>
void
test()
{
...
}

int main()
{
...
    test<0, false>();
```

Did you forget to update the template arguments?

masterleinad commented:
Would you mind just pushing the tests anyway so it's easier for someone else to have a look and we can also see how the CI responds (in case it reports something at compile-time).

tjhei commented:
Does the compiler need explicit instantiations of the helper functions?

YiminJin commented:
> Would you mind just pushing the tests anyway so it's easier for someone else to have a look and we can also see how the CI responds (in case it reports something at compile-time).

Sure. I have pushed the new changes, include the tests, the asserts and the renaming of copy_view_1d. Let's see what the checkers say.

YiminJin commented:
> Does the compiler need explicit instantiations of the helper functions?

I tried to add
```
template void test<3, 4, 0, false>();
```
after the instantiation of `test()` in file evaluate_2d_shape_3x4.cc, but it does not work. The same error occurs.

YiminJin commented:
It seems that the test file did not encounter runtime errors on the checkers... So the problem might be in my configuration of CUDA or Kokkos. I did encounter a lot of problems when installing CUDA on my machines. In fact, I had to hack into the source codes of gcc and nvcc (following the instructions of CUDA forum) in order to install deal.II on my laptop.

Anyway, good news for me. I have updated the outputs of the modified tests and squashed the commits again. Hope everything goes well.

masterleinad commented:
> It seems that the test file did not encounter runtime errors on the checkers... So the problem might be in my configuration of CUDA or Kokkos. I did encounter a lot of problems when installing CUDA on my machines. In fact, I had to hack into the source codes of gcc and nvcc (following the instructions of CUDA forum) in order to install deal.II on my laptop.

The CI is not actually running on any GPU so it might be good to check manually. At least, runninng on CPUs seems to work.

masterleinad commented:
I'm also seeing
```
7028: Test command: /soft/spack/base/0.8.1/install/linux-sles15-x86_64/gcc-12.3.0/cmake-3.27.9-cs3caqvuqem4vxdrxyrkm6ehddunt75s/bin/cmake "-DTRGT=matrix_free_kokkos.evaluate_2d_shape.debug.test" "-DTEST=matrix_free_kokkos/evaluate_2d_shape.debug" "-DEXPECT=PASSED" "-DBINARY_DIR=/home/darndt/dealii/build/tests/matrix_free_kokkos" "-P" "/home/darndt/dealii/build/share/deal.II/scripts/run_test.cmake"
7028: Working Directory: /home/darndt/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.debug/serial
7028: Test timeout computed to be: 600
7028: /home/darndt/dealii/build/share/deal.II/scripts/run_test.sh: line 34: 226232 Aborted                 (core dumped) "$@" > stdout 2>&1
7028: gmake[3]: *** [CMakeFiles/matrix_free_kokkos.evaluate_2d_shape.debug.test.dir/build.make:78: evaluate_2d_shape.debug/serial/output] Error 1
7028: gmake[2]: *** [CMakeFiles/Makefile2:465: CMakeFiles/matrix_free_kokkos.evaluate_2d_shape.debug.test.dir/all] Error 2
7028: gmake[1]: *** [CMakeFiles/Makefile2:472: CMakeFiles/matrix_free_kokkos.evaluate_2d_shape.debug.test.dir/rule] Error 2
7028: gmake: *** [Makefile:277: matrix_free_kokkos.evaluate_2d_shape.debug.test] Error 2
7028: Test matrix_free_kokkos/evaluate_2d_shape.debug: RUN
7028: ===============================   OUTPUT BEGIN  ===============================
7028: [ 33%] Building CXX object CMakeFiles/matrix_free_kokkos.evaluate_2d_shape.debug.dir/evaluate_2d_shape.cc.o
7028: [ 66%] Linking CXX executable evaluate_2d_shape.debug/evaluate_2d_shape.debug
7028: [ 66%] Built target matrix_free_kokkos.evaluate_2d_shape.debug
7028: [100%] Normalizing test output file /home/darndt/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.debug/serial/output
7028: matrix_free_kokkos/evaluate_2d_shape.debug: BUILD successful.
7028: matrix_free_kokkos/evaluate_2d_shape.debug: RUN failed. ------ Return code 134
7028: matrix_free_kokkos/evaluate_2d_shape.debug: RUN failed. ------ Result: /home/darndt/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.debug/serial/failing_output
7028: matrix_free_kokkos/evaluate_2d_shape.debug: RUN failed. ------ Partial output:
7028: JobId x3003c0s19b1n0 Tue Mar  4 23:26:49 2025
7028: DEAL:values::Test 4 x 4
7028: DEAL:values::Errors no transpose: -1.11022e-16 5.55112e-17 1.11022e-16 2.22045e-16 -1.38778e-17 1.73472e-17 5.55112e-17 1.24900e-16 -4.16334e-17 6.93889e-18 -5.55112e-17 -1.59595e-16 -2.22045e-16 2.77556e-17 -5.55112e-17 0.00000 
7028: DEAL:values::Errors transpose:    -4.44089e-16 1.11022e-16 0.00000 -8.88178e-16 -2.22045e-16 6.93889e-17 8.32667e-17 -1.11022e-16 0.00000 -2.77556e-17 1.11022e-16 4.44089e-16 0.00000 0.00000 0.00000 0.00000 
7028: DEAL:values::Test 3 x 3
7028: DEAL:values::Errors no transpose: 0.00000 -6.93889e-18 1.73472e-18 0.00000 -2.77556e-17 0.00000 1.73472e-18 0.00000 0.00000 
7028: DEAL:values::Errors transpose:    0.00000 0.00000 -1.38778e-17 0.00000 0.00000 0.00000 3.46945e-18 0.00000 -1.04083e-17 
7028: DEAL:values::Test 3 x 4
7028: 
7028: matrix_free_kokkos/evaluate_2d_shape.debug: RUN failed. ------ Additional output on stdout/stderr:
7028: 
7028: terminate called after throwing an instance of 'std::runtime_error'
7028:   what():  cudaFuncGetAttributes(&attr, func) error( cudaErrorInvalidDeviceFunction): invalid device function /home/darndt/kokkos-install/include/Cuda/Kokkos_Cuda_KernelLaunch.hpp:140
7028: 
7028: 
7028: matrix_free_kokkos/evaluate_2d_shape.debug: ******    RUN failed    *******
7028: 
7028: ===============================    OUTPUT END   ===============================
7028: Expected stage PASSED - aborting
7028: CMake Error at /home/darndt/dealii/build/share/deal.II/scripts/run_test.cmake:115 (message):
7028:   *** abort
7028: 
7028: 
 5/32 Test #7028: matrix_free_kokkos/evaluate_2d_shape.debug .............................***Failed   41.39 sec
test 7029
      Start 7029: matrix_free_kokkos/evaluate_2d_shape.release

7029: Test command: /soft/spack/base/0.8.1/install/linux-sles15-x86_64/gcc-12.3.0/cmake-3.27.9-cs3caqvuqem4vxdrxyrkm6ehddunt75s/bin/cmake "-DTRGT=matrix_free_kokkos.evaluate_2d_shape.release.test" "-DTEST=matrix_free_kokkos/evaluate_2d_shape.release" "-DEXPECT=PASSED" "-DBINARY_DIR=/home/darndt/dealii/build/tests/matrix_free_kokkos" "-P" "/home/darndt/dealii/build/share/deal.II/scripts/run_test.cmake"
7029: Working Directory: /home/darndt/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.release/serial
7029: Test timeout computed to be: 600
7029: /home/darndt/dealii/build/share/deal.II/scripts/run_test.sh: line 34: 226725 Aborted                 (core dumped) "$@" > stdout 2>&1
7029: gmake[3]: *** [CMakeFiles/matrix_free_kokkos.evaluate_2d_shape.release.test.dir/build.make:78: evaluate_2d_shape.release/serial/output] Error 1
7029: gmake[2]: *** [CMakeFiles/Makefile2:517: CMakeFiles/matrix_free_kokkos.evaluate_2d_shape.release.test.dir/all] Error 2
7029: gmake[1]: *** [CMakeFiles/Makefile2:524: CMakeFiles/matrix_free_kokkos.evaluate_2d_shape.release.test.dir/rule] Error 2
7029: gmake: *** [Makefile:303: matrix_free_kokkos.evaluate_2d_shape.release.test] Error 2
7029: Test matrix_free_kokkos/evaluate_2d_shape.release: RUN
7029: ===============================   OUTPUT BEGIN  ===============================
7029: [ 33%] Building CXX object CMakeFiles/matrix_free_kokkos.evaluate_2d_shape.release.dir/evaluate_2d_shape.cc.o
7029: [ 66%] Linking CXX executable evaluate_2d_shape.release/evaluate_2d_shape.release
7029: [ 66%] Built target matrix_free_kokkos.evaluate_2d_shape.release
7029: [100%] Normalizing test output file /home/darndt/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.release/serial/output
7029: matrix_free_kokkos/evaluate_2d_shape.release: BUILD successful.
7029: matrix_free_kokkos/evaluate_2d_shape.release: RUN failed. ------ Return code 134
7029: matrix_free_kokkos/evaluate_2d_shape.release: RUN failed. ------ Result: /home/darndt/dealii/build/tests/matrix_free_kokkos/evaluate_2d_shape.release/serial/failing_output
7029: matrix_free_kokkos/evaluate_2d_shape.release: RUN failed. ------ Partial output:
7029: JobId x3003c0s19b1n0 Tue Mar  4 23:27:31 2025
7029: DEAL:values::Test 4 x 4
7029: DEAL:values::Errors no transpose: -1.11022e-16 5.55112e-17 1.11022e-16 2.22045e-16 -1.38778e-17 1.73472e-17 5.55112e-17 1.24900e-16 -4.16334e-17 6.93889e-18 -5.55112e-17 -1.59595e-16 -2.22045e-16 2.77556e-17 -5.55112e-17 0.00000 
7029: DEAL:values::Errors transpose:    -4.44089e-16 1.11022e-16 0.00000 -8.88178e-16 -2.22045e-16 6.93889e-17 8.32667e-17 -1.11022e-16 0.00000 -2.77556e-17 1.11022e-16 4.44089e-16 0.00000 0.00000 0.00000 0.00000 
7029: DEAL:values::Test 3 x 3
7029: DEAL:values::Errors no transpose: 0.00000 -6.93889e-18 1.73472e-18 0.00000 -2.77556e-17 0.00000 1.73472e-18 0.00000 0.00000 
7029: DEAL:values::Errors transpose:    0.00000 0.00000 -1.38778e-17 0.00000 0.00000 0.00000 3.46945e-18 0.00000 -1.04083e-17 
7029: DEAL:values::Test 3 x 4
7029: 
7029: matrix_free_kokkos/evaluate_2d_shape.release: RUN failed. ------ Additional output on stdout/stderr:
7029: 
7029: terminate called after throwing an instance of 'std::runtime_error'
7029:   what():  cudaFuncGetAttributes(&attr, func) error( cudaErrorInvalidDeviceFunction): invalid device function /home/darndt/kokkos-install/include/Cuda/Kokkos_Cuda_KernelLaunch.hpp:140
7029: 
7029: 
7029: matrix_free_kokkos/evaluate_2d_shape.release: ******    RUN failed    *******
7029: 
7029: ===============================    OUTPUT END   ===============================
7029: Expected stage PASSED - aborting
7029: CMake Error at /home/darndt/dealii/build/share/deal.II/scripts/run_test.cmake:115 (message):
7029:   *** abort
7029: 
7029: 
```
so we need to figure this out.

masterleinad commented:
Using an explicit functor instead of a lambda is fixing the issue. I pushed a corresponding commit.

kronbichler reviewed (approved):
I think this is good progress, thanks a lot for pushing this forward! I have a few modifications around locally, which I will later adapt to this new code, and thus further improve upon this with that.

tjhei commented:
awesome, thank you @YiminJin and @masterleinad !

YiminJin commented:
I am sorry about this. I have noticed the error message `Memory access fault by GPU node-2`, yet I cannot reproduce it because I do not have a server account with GPU in China. The only thing I can do here is to list the functions called by step-64:
`Portable::FEEvaluation<3,3,4,1>::evaluate()` and `Portable::FEEvaluation<3,3,4,1>::integrate()`
(include/dealii/matrix_free/portable_fe_evaluation.h)
`Portable::internal::FEEvaluationImplTransformToCollocation<3,3>::evaluate()` and
`Portable::internal::FEEvaluationImplTransformToCollocation<3,3>::integrate()` 
(include/dealii/matrix_free/portable_evaluation_kernels.h)
`Portable::internal::EvaluatorTensorProduct<EvaluatorVariant::evaluate_general,3,4,4>::values<direction,dof_to_quad,add,in_place>()` and
`Portable::internal::EvaluatorTensorProduct<EvaluatorVariant::evaluate_general,3,4,4>::co_gradients<direction,dof_to_quad,add,in_place>()`
(include/dealii/matrix_free/portable_tensor_product_kernels.h)
`Portable::internal::apply<3,4,4,Number,direction,contract_over_rows,add>::apply()`
(include/dealii/matrix_free/portable_tensor_product_kernels.h)

I hope this helps.
